File: Paket/Inti/lib/Nilai/Uang.dart
import 'package:decimal/decimal.dart';

import 'ModePembulatan.dart';

/// Nilai uang Rupiah dengan presisi tetap 2 desimal (PRD §8 F-07, §15.1, CLAUDE.md #7).
///
/// Padanan `App\Domain\Bersama\Nilai\Uang` di Backend. Tidak pernah memakai tipe pecahan biner.
/// Nilai dengan lebih dari 2 desimal ditolak, kecuali lewat operasi yang menyebut mode pembulatan.
final class Uang implements Comparable<Uang> {
  const Uang._(this._nilai);

  static const int skala = 2;

  final Decimal _nilai;

  /// Dari string desimal, misal `"15000"` atau `"18000.5"`.
  static Uang Dari(String nilai) => DariDesimal(Decimal.parse(nilai));

  /// Dari Rupiah bulat, misal `15000`.
  static Uang DariBulat(int rupiah) => Uang._(Decimal.fromInt(rupiah));

  static Uang DariDesimal(Decimal nilai) {
    if (nilai.scale > skala) {
      throw ArgumentError.value(nilai.toString(), 'nilai', 'Uang maksimal $skala desimal');
    }
    return Uang._(nilai);
  }

  static Uang Nol() => Uang._(Decimal.zero);

  Uang Tambah(Uang lain) => Uang._(_nilai + lain._nilai);

  Uang Kurangi(Uang lain) => Uang._(_nilai - lain._nilai);

  /// Mengalikan dengan jumlah/tarif lalu membulatkan ke 2 desimal dengan mode yang disebut eksplisit.
  Uang Kali(Decimal faktor, {ModePembulatan mode = ModePembulatan.setengahMenjauhiNol}) =>
      Uang._(BulatkanKeSkala(_nilai * faktor, skala, mode));

  /// Pembulatan ke kelipatan tertentu, misal pembulatan tunai ke Rp 100 (PRD Lampiran D).
  Uang BulatkanKeKelipatan(int kelipatan, ModePembulatan mode) {
    if (kelipatan <= 0) {
      throw ArgumentError.value(kelipatan, 'kelipatan', 'Kelipatan pembulatan harus lebih dari 0');
    }
    final sen = _nilai.shift(skala).toBigInt();
    final penyebut = BigInt.from(kelipatan) * BigInt.from(100);
    return Uang._(Decimal.fromBigInt(BagiBulat(sen, penyebut, mode) * BigInt.from(kelipatan)));
  }

  /// Hasil selalu -1, 0, atau 1 (sama dengan `BigDecimal::compareTo` di Backend).
  int Bandingkan(Uang lain) => _nilai.compareTo(lain._nilai).sign;

  bool SamaDengan(Uang lain) => _nilai == lain._nilai;

  bool BernilaiNol() => _nilai == Decimal.zero;

  bool BernilaiNegatif() => _nilai < Decimal.zero;

  Decimal KeDesimal() => _nilai;

  /// Format penyimpanan & API: string desimal, misal `"15000.00"`.
  String KeString() => _nilai.toStringAsFixed(skala);

  /// Format tampilan Indonesia, misal `"Rp 1.250.000"` atau `"−Rp 6.000"`.
  String FormatRupiah() {
    final mutlak = _nilai.abs();
    final bulat = mutlak.truncate();
    final sen = (mutlak - bulat).shift(skala).toBigInt();
    final digit = bulat.toBigInt().toString();
    final kelompok = StringBuffer();
    for (var indeks = 0; indeks < digit.length; indeks++) {
      if (indeks > 0 && (digit.length - indeks) % 3 == 0) {
        kelompok.write('.');
      }
      kelompok.write(digit[indeks]);
    }
    final teksSen = sen == BigInt.zero ? '' : ',${sen.toString().padLeft(2, '0')}';
    return '${BernilaiNegatif() ? '−' : ''}Rp $kelompok$teksSen';
  }

  String toJson() => KeString();

  @override
  int compareTo(Uang other) => Bandingkan(other);

  @override
  bool operator ==(Object other) => other is Uang && SamaDengan(other);

  @override
  int get hashCode => _nilai.hashCode;

  @override
  String toString() => KeString();
}

File: Aplikasi/Web/app/Domain/Bersama/Nilai/Uang.php
<?php

declare(strict_types=1);

namespace App\Domain\Bersama\Nilai;

use Brick\Math\BigDecimal;
use Brick\Math\BigNumber;
use Brick\Math\RoundingMode;
use InvalidArgumentException;
use JsonSerializable;
use Stringable;

/**
 * Nilai uang Rupiah dengan presisi tetap 2 desimal (PRD §8 F-07, §15.1, CLAUDE.md #7).
 *
 * Tidak pernah memakai float. Masukan hanya int, string desimal, atau BigNumber.
 * Nilai dengan lebih dari 2 desimal ditolak, kecuali lewat operasi yang menyebut mode pembulatan.
 */
final readonly class Uang implements JsonSerializable, Stringable
{
    public const SKALA = 2;

    private function __construct(private BigDecimal $nilai) {}

    public static function Dari(BigNumber|int|string $nilai): self
    {
        $desimal = BigDecimal::of($nilai);

        if ($desimal->getScale() > self::SKALA && ! $desimal->toScale(self::SKALA, RoundingMode::Down)->isEqualTo($desimal)) {
            throw new InvalidArgumentException('Uang maksimal '.self::SKALA." desimal: {$nilai}");
        }

        return new self($desimal->toScale(self::SKALA));
    }

    public static function Nol(): self
    {
        return new self(BigDecimal::zero()->toScale(self::SKALA));
    }

    public function Tambah(self $lain): self
    {
        return new self($this->nilai->plus($lain->nilai));
    }

    public function Kurangi(self $lain): self
    {
        return new self($this->nilai->minus($lain->nilai));
    }

    /**
     * Mengalikan dengan jumlah/tarif lalu membulatkan ke 2 desimal dengan mode yang disebut eksplisit.
     */
    public function Kali(BigNumber|int|string $faktor, RoundingMode $mode = RoundingMode::HalfUp): self
    {
        return new self($this->nilai->multipliedBy($faktor)->toScale(self::SKALA, $mode));
    }

    /**
     * Pembulatan ke kelipatan tertentu, misal pembulatan tunai ke Rp 100 (PRD Lampiran D).
     */
    public function BulatkanKeKelipatan(int $kelipatan, RoundingMode $mode): self
    {
        if ($kelipatan <= 0) {
            throw new InvalidArgumentException('Kelipatan pembulatan harus lebih dari 0.');
        }

        $dibulatkan = $this->nilai->dividedBy($kelipatan, 0, $mode)->multipliedBy($kelipatan);

        return new self($dibulatkan->toScale(self::SKALA));
    }

    public function Bandingkan(self $lain): int
    {
        return $this->nilai->compareTo($lain->nilai);
    }

    public function SamaDengan(self $lain): bool
    {
        return $this->nilai->isEqualTo($lain->nilai);
    }

    public function BernilaiNol(): bool
    {
        return $this->nilai->isZero();
    }

    public function BernilaiNegatif(): bool
    {
        return $this->nilai->isNegative();
    }

    /** Format penyimpanan & API: string desimal, misal "15000.00". */
    public function KeString(): string
    {
        return (string) $this->nilai;
    }

    /** Format tampilan Indonesia, misal "Rp 1.250.000" atau "−Rp 6.000". */
    public function FormatRupiah(): string
    {
        $mutlak = $this->nilai->abs();
        $bulat = $mutlak->toScale(0, RoundingMode::Down);
        $sen = (string) $mutlak->minus($bulat)->multipliedBy(100)->toScale(0);
        $teks = 'Rp '.number_format((int) (string) $bulat, 0, ',', '.');

        if ($sen !== '0') {
            $teks .= ','.str_pad($sen, 2, '0', STR_PAD_LEFT);
        }

        return ($this->nilai->isNegative() ? '−' : '').$teks;
    }

    public function jsonSerialize(): string
    {
        return $this->KeString();
    }

    public function __toString(): string
    {
        return $this->KeString();
    }
}
